ga-migration script part II

mvidalgarcia opened the pull request.

avivace committed:
ga-migration: add pypi-publish.yml if needed

- Add pypi-publish.yml only if the publish step was mentioned in the travis.yml file

- Split pipeline and utility methods

mvidalgarcia committed:
ga-migration: add logic to build from templates

- Identify repository services.
- Generate and copy `test.yml` & `run-tests.sh` using Jinja2.

closes #3

diegodelemos commented on automation_tools/scripts/ga-migration/templates/services/tests.yml:
```diff
@@ -0,0 +1,144 @@
+          {{ "mq-service: [rabbitmq, redis]" if mq }}
+          {{ "search-service: [elasticsearch6, elasticsearch7]" if search }}
+
+          exclude:
```
Major: we should remember to come up with the list mentioned in https://github.com/inveniosoftware/cookiecutter-invenio-module/pull/141#discussion_r534029729, to we have a list of excluded builds by default so when we make all the PRs we do not consume our quota.

diegodelemos commented on automation_tools/scripts/ga-migration/utils.py:
```diff
@@ -5,13 +5,75 @@
+
+    search_terms = {
+        'postgres': 'db',
+        '- redis': 'cache',
```
Why not `redis-service`? e.g. https://github.com/inveniosoftware/invenio-cache/blob/50879bbf441718eebb81a0356852b76590166824/.travis.yml#L24

mvidalgarcia replied:
I found **only** `- redis` in other repos like [invenio-indexer](https://github.com/inveniosoftware/invenio-indexer/blob/0b50ed3f8a3ec683ae6a1e697cd04222a0d8c569/.travis.yml#L25) so in the end I think lowest common denominator is that. What do you think?

diegodelemos replied:
Good enough, thanks!

avivace committed:
ga-migrations: initial gitflow implemenation

- Settings things up for implementing the git flow

- Walk past commits, stub PR and commit steps

diegodelemos commented:
With:
```console
automation-tools $ git log -1 --oneline
18915be (HEAD -> ga-migration, upstream/ga-migration) ga-migration: add logic to build from templates
invenio-stats $ git log -1 --oneline
f73609b (HEAD -> master, tag: v1.0.0a18, upstream/master, origin/master, origin/HEAD) release: v1.0.0a18
```
I get the following error:
```console
$ python main.py --targetpath ~/project/invenio/src/invenio-stats # also tried on invenio-queues
INFO:root:TASK: Building and copying tests.yml template...
Traceback (most recent call last):
  File "main.py", line 141, in <module>
    pipeline()
  File "/Users/rodrigdi/.pyenv/versions/gh-migrate-helper/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/rodrigdi/.pyenv/versions/gh-migrate-helper/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/rodrigdi/.pyenv/versions/gh-migrate-helper/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/rodrigdi/.pyenv/versions/gh-migrate-helper/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "main.py", line 134, in pipeline
    migrate_repo(targetpath)
  File "main.py", line 71, in migrate_repo
    build_template(
  File "/Users/rodrigdi/project/invenio/src/automation-tools/automation_tools/scripts/ga-migration/utils.py", line 63, in build_template
    content = render_template(
  File "/Users/rodrigdi/project/invenio/src/automation-tools/automation_tools/scripts/ga-migration/utils.py", line 73, in render_template
    template = JinjaEnv.get_template(template_file)
  File "/Users/rodrigdi/.pyenv/versions/gh-migrate-helper/lib/python3.8/site-packages/jinja2/environment.py", line 883, in get_template
    return self._load_template(name, self.make_globals(globals))
  File "/Users/rodrigdi/.pyenv/versions/gh-migrate-helper/lib/python3.8/site-packages/jinja2/environment.py", line 857, in _load_template
    template = self.loader.load(self, name, globals)
  File "/Users/rodrigdi/.pyenv/versions/gh-migrate-helper/lib/python3.8/site-packages/jinja2/loaders.py", line 115, in load
    source, filename, uptodate = self.get_source(environment, name)
  File "/Users/rodrigdi/.pyenv/versions/gh-migrate-helper/lib/python3.8/site-packages/jinja2/loaders.py", line 197, in get_source
    raise TemplateNotFound(template)
jinja2.exceptions.TemplateNotFound: services/tests.yml
```

mvidalgarcia committed:
ga-migration: re-check already migrated repos

mvidalgarcia commented:
@diegodelemos It should be fixed by [this change](https://github.com/inveniosoftware/automation-tools/pull/9/commits/f3d5c427e4da6def6058768c628525c712746475#diff-334e9270f835f084ff3ae3f38cb1c0274ccf1f2301a480a3afc47481f260a30fR19), thanks for reporting 👍
Also, already migrated repos should be taken into account now.

avivace committed:
ga-migration: gitflow auth and fixes

- GH auth

- Implemented basic git flow

- Handle .travis.yml present in the most recent commit

- Upgrade Sphinx 1 to 3 (#10)

- Skip replace_list if file is not found

- Delete baks afterwards

diegodelemos commented on automation_tools/scripts/ga-migration/templates/services/run-tests.sh:
```diff
@@ -0,0 +1,30 @@
+#
+# Copyright (C) 2020 CERN.
+#
+# Invenio-Records-Resources is free software; you can redistribute it and/or modify
```
Could we change this to the generic Invenio license header? e.g. https://github.com/inveniosoftware/automation-tools/pull/9/files#diff-707cee7a3f79b3ef314738bb5c26a4fb63dc739a84046a96c0eaeed37374a374

mvidalgarcia replied:
Yes, I have it locally, pushing soon.

mvidalgarcia replied:
Pushed with some other improvements:

```
- Adapt to latest `docker-services-cli` version
- Fix various issues with Jinja white-spaces
- Emulate join function to generate EXTRAS
```

mvidalgarcia committed:
ga-migration: multiple improvements

- Adapt to latest `docker-services-cli` version
- Fix various issues with Jinja white-spaces
- Emulate join function to generate EXTRAS

avivace committed:
ga-migration: update docs and minor fixes

- Update requirements.txt

- Update CLI interface and README

- Skip replace_list if the regex match fails

- Skip pypi-publish task if it's not mentioned in .travis.yml

- Add deleted bulk applying instrauctions

mvidalgarcia commented on automation_tools/scripts/ga-migration/README.md:
```diff
@@ -8,12 +8,34 @@ $ source ~/.virtualenvs/ga-migration/bin/activate
+## Full pipeline
 
-# Migrate all repos on config.py
-$ python main.py
```
Why have you deleted this option? It's useful to bulk-migrate repos

avivace replied:
Woops

lnielsen reviewed (commented):
Comment: Just noticed the commit history - please try to keep logical commits instead of many smaller commits. It get's very difficult both to review, and go back and find bugs when there's a lot of small commits.
